code/modules/mob/living/blood.dm

```diff
@@ -138,17 +138,27 @@
 
 	if(iscarbon(AM))
 		var/mob/living/carbon/C = AM
-		if(blood_id == C.get_blood_id())//both mobs have the same blood substance
-			if(blood_id == "blood") //normal blood
-				if(blood_data["viruses"])
-					for(var/thing in blood_data["viruses"])
-						var/datum/disease/D = thing
-						if((D.spread_flags & DISEASE_SPREAD_SPECIAL) || (D.spread_flags & DISEASE_SPREAD_NON_CONTAGIOUS))
-							continue
-						C.ForceContractDisease(D)
-				if(!(blood_data["blood_type"] in get_safe_blood(C.dna.blood_type)))
-					C.reagents.add_reagent("toxin", amount * 0.5)
-					return 1
+		for(var/datum/reagent/blood/bluhduh in GLOB.blood_types[blood_id])
+			if(blood_id == C.get_blood_id())//both mobs have the same blood substance
+				if(bluhduh) //actual blood reagent
+					if(blood_data["viruses"])
+						for(var/thing in blood_data["viruses"])
+							var/datum/disease/D = thing
+							if((D.spread_flags & DISEASE_SPREAD_SPECIAL) || (D.spread_flags & DISEASE_SPREAD_NON_CONTAGIOUS))
+								continue
+							C.ForceContractDisease(D)
+					if(!(blood_data["blood_type"] in get_safe_blood(C.dna.blood_type)))
+						if((blood_data["blood_type"] == "GEL") && (C.dna.species.exotic_blood != "jellyblood"))
+							C.reagents.add_reagent("bonehurtingjuice", amount * 1.5) //filthy xenos bloooood
+						else if((blood_data["blood_type"] == "HF") && (C.dna.species.exotic_blood != "oilblood"))
+							C.reagents.add_reagent("bonehurtingjuice", amount * 1)	//don't fucking put oil in people
+						else if((blood_data["blood_type"] == "X*") && (C.dna.species.exotic_blood != "xenoblood"))
+							C.reagents.add_reagent("sacid", amount * 1.5) //acid blooood
+						else
+							C.reagents.add_reagent("bonehurtingjuice", amount * 0.5)
+					else
+						C.blood_volume = min(C.blood_volume + round(amount, 0.1), BLOOD_VOLUME_MAXIMUM)
+						return 1
```

Comment:
```suggestion
					return 1
```
stopping the target's blood volume from being resotred from incompatible bloods.

Also, this else if list is sucky, but I'm getting tired of reviewing this.